recipes/terraclimate/recipe.py

```diff
@@ -0,0 +1,103 @@
+from pangeo_forge_recipes.recipes import XarrayZarrRecipe
+from pangeo_forge_recipes.patterns import FilePattern, ConcatDim, MergeDim
+import xarray as xr
+
+target_chunks = {"lat": 1024, "lon": 1024, "time": 12}
+# only do two years to keep the example small; it's still big!
+years = list(range(1958, 1960))
+variables = [
+    "aet",
+    "def",
+    "pet",
+    "ppt",
+    "q",
+    "soil",
+    "srad",
+    "swe",
+    "tmax",
+    "tmin",
+    "vap",
+    "ws",
+    "vpd",
+    "PDSI",
+]
+
+def make_filename(variable, time):
+    return f"http://thredds.northwestknowledge.net:8080/thredds/fileServer/TERRACLIMATE_ALL/data/TerraClimate_{variable}_{time}.nc"
+
+pattern = FilePattern(
+    make_filename,
+    ConcatDim(name="time", keys=years),
+    MergeDim(name="variable", keys=variables)
+)
+
+rename_vars = {'PDSI': 'pdsi'}
+
+mask_opts = {
+    "PDSI": ("lt", 10),
+    "aet": ("lt", 32767),
+    "def": ("lt", 32767),
+    "pet": ("lt", 32767),
+    "ppt": ("lt", 32767),
+    "ppt_station_influence": None,
+    "q": ("lt", 2147483647),
+    "soil": ("lt", 32767),
+    "srad": ("lt", 32767),
+    "swe": ("lt", 10000),
+    "tmax": ("lt", 200),
+    "tmax_station_influence": None,
+    "tmin": ("lt", 200),
+    "tmin_station_influence": None,
+    "vap": ("lt", 300),
+    "vap_station_influence": None,
+    "vpd": ("lt", 300),
+    "ws": ("lt", 200),
+}
+
+def apply_mask(key, da):
+    """helper function to mask DataArrays based on a threshold value"""
+    if mask_opts.get(key, None):
+        op, val = mask_opts[key]
+        if op == "lt":
+            da = da.where(da < val)
+        elif op == "neq":
+            da = da.where(da != val)
+    return da
+
+
+def preproc(ds):
+    """custom preprocessing function for terraclimate data"""
+    
```

Comment:
```suggestion
    from builtins import list

```

Reply:
It seems utterly bizarre to me that beam would expect you to do `from builtins import list`.

Reply:
Agreed, this is very strange. Though https://cloud.google.com/dataflow/docs/resources/faq#how_do_i_handle_nameerrors does say...

> If the error is regarding a module that should be available in the Python distribution, you can solve this by importing the module locally, where it is used.
>
> For example, instead of:
> 
> ```
> import re
> …
> def myfunc():
>   # use re module
> ```
> 
> use:
> 
> ```
> def myfunc():
>   import re
>   # use re module
>  ```

🤷 It's cheap enough to try this, though even if it does work, I'm not of the view that it's the preferred long-term solution...

Reply:
Going to commit this for experimental (not stylistic) value ...